src/Features/Core/Portable/CodeRefactorings/AbstractRefactoringHelpersService.cs

```diff
@@ -0,0 +1,176 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings
+{
+    internal abstract class AbstractRefactoringHelpersService : IRefactoringHelpersService
+    {
+        public async Task<TSyntaxNode> TryGetSelectedNodeAsync<TSyntaxNode>(
+            Document document, TextSpan selection, CancellationToken cancellationToken) where TSyntaxNode : SyntaxNode
+        {
+            return await TryGetSelectedNodeAsync(document, selection, n => n is TSyntaxNode, cancellationToken).ConfigureAwait(false) as TSyntaxNode;
```

Comment:
see my previous comment about how i think this predicate could be entirely removed without any real loss in expressiveness.

Reply:
Let's leave it there for now (I can imagine that some more complex scenarios (e.g. I want a node of type X whose parent is Y) might come up. We can always remove it later (will write it down to not to remember it).

Reply:
Hrmm... those extra scenarios feel like they would be fine for the caller to encode themselves though.  i.e. i try to get the node back.  if it has the wrong parent, then it's not a good node.  etc. etc.

Reply:
I can completely understand the desire to have as simple API surface as possible but I think having accessible overloads for potentially more complex situation isn't a bad idea either. If we don't see uses for them by the time more Refactorings are moved we can always remove them.

Reply:
Ok, what about this: let's remove it from the service's API surface but leave the infrastructure for it in the helpers themselves. If its not required then we can simply refactor it (I'll add note to not to forget about it). And if it turns out to be useful we can re-introduce it.

Reply:
BTW: Is this really more readable than a single lambda that obviously tests against two types? 

```csharp

            var memberDecl = await refactoringHelperService.TryGetSelectedNodeAsync<MemberDeclarationSyntax>(document, span, cancellationToken).ConfigureAwait(false);
            if (memberDecl != default)
            {
                return memberDecl;
            }

            var varDecl = await refactoringHelperService.TryGetSelectedNodeAsync<VariableDeclaratorSyntax>(document, span, cancellationToken).ConfigureAwait(false);
            return varDecl;
```

// I wish I could more easily abstract over set of types but...

Reply:
And in `LambdaSimplifierCodeRefactoringProvider` we'll have to go from:
```csharp
            var lambda = semanticDocument.Root.FindToken(textSpan.Start).GetAncestor(n =>
                n is SimpleLambdaExpressionSyntax || n is ParenthesizedLambdaExpressionSyntax);
            if (lambda == null || !lambda.Span.IntersectsWith(textSpan.Start))
``` 

to: 
```csharp
            var refactoringHelperService = document.GetLanguageService<IRefactoringHelpersService>();

            SyntaxNode lambda = await refactoringHelperService.TryGetSelectedNodeAsync<SimpleLambdaExpressionSyntax>(document, context.Span, cancellationToken).ConfigureAwait(false);
            if (lambda == null)
            {
                lambda = await refactoringHelperService.TryGetSelectedNodeAsync<ParenthesizedLambdaExpressionSyntax>(document, context.Span, cancellationToken).ConfigureAwait(false);
            }
            if (lambda == null)
```

PS: Actually that one will require more changes because `).GetAncestor(n =>
                n is SimpleLambdaExpressionSyntax || n is ParenthesizedLambdaExpressionSyntax)` traverses the Node tree upwards (not only on edges) but let's handle that in next PR.

Reply:
Similarly `AbstractUseTypeCodeRefactoringProvider` handles: 
```csharp
var token = root.FindToken(textSpan.Start);
            return token.Parent?.FirstAncestorOrSelf<SyntaxNode>(
                a => a.IsKind(SyntaxKind.DeclarationExpression, SyntaxKind.VariableDeclaration, SyntaxKind.ForEachStatement));
```

which would require 3 calls...

Reply:
> but I think having accessible overloads for potentially more complex situation isn't a bad idea either. 

That's premature speculation :D

> And in LambdaSimplifierCodeRefactoringProvider we'll have to go from:

Or just go to: `TryGetSelectedNodeAsync<LambdaExpressionSyntax>(...)`

Reply:
> BTW: Is this really more readable than a single lambda that obviously tests against two types?

To me, yes.  Each expresses the intent well, and would be easy to comment and explain.  i.e "// we look for the member decl form to handle thsee cases"  and then "// but we also check for the var-decl form to ensure we get these sorts of selections".  

Brevity is not necessarily clarity.

Reply:
I'm ok with going in the direction you wanted and have multiple calls for these situations but I want to make sure the context is clear. 

Also we already have methods taking `Predicate<>` in the codebase so it's not even a new pattern.

Reply:
> Ok, what about this: let's remove it from the service's API surface but leave the infrastructure for it in the helpers themselves.

Of course.  that's totally fine with me.  I was referring the public surface area i would expect clients to use.

Note: part of this is driven by a goal of makign these avialably publicly.  That means definitely keeping a mind toward simplicity and clarity over making things unnecessarily complex or trying to overly optimize for non-hotpath scenarios.

Reply:
> which would require 3 calls...

That's 100% ok with me :)

Reply:
Note: if you really want, you could have an extension like this:

```c#
interface IRefactoringHelperServiceExtensions 
{
     public async Task<SyntaxNode> TryGetSelectedNode<TNode1, TNode2, TNode3>(this IRefactoringHelperService, Document document, TextSpan span, CancellationToken)
    {
          // impl that tries each of the TNodes
    }
}
```

But i'm highly skeptical this is needed, or that it would be undesirable to just have the caller make 2-3 calls themselves.

Reply:
> Also we already have methods taking Predicate<> in the codebase so it's not even a new pattern.

In general, the predice form was necessary because it's not actually correct to make multiple calls (For some operations).  Consider "GetAncestorOrThis".  If you're trying to find the the *first* matching ancestor, then you need the predicate.  Otherwise, say you do `GetAncestorOrThis<XType>(...)` and `GetAncestorOrThis<YType>(...)`.  The first might find a *higher* node than the second, and ther'es no easy way to see which was 'first'.  Hence why some of those calls take a predicate.

This isn't really necessary in your case because because for the selection case, we don't have that same issue as higher/lower nodes will likely be knocked out by virtue of not matching the selection properly.

Reply:
> Brevity is not necessarily clarity. 
good point

And generally I've been convinced. Added one comment & will merge after that.